### ISSUE TITLE
Script only partially works with Userscripts

### ISSUE THREAD

**Issue by hopkapi** [quoid/userscripts#383]
_System Information:_
> macOS or iOS version: 13.1 Beta (22C5050e)
> Userscripts version: 4.4.3 (65)
> Safari version: 16.2 (18614.3.6.11.2)
> Is this issue related to script injection? I think so (?)
> Did the test script (pasted above) successfully run on your machine? Yes

I can't seem to get [this script](https://gist.github.com/busybox11/53c76f57a577a47a19fab649a76f18e3) to work fully with Userscripts. It's meant to change the normal Verified icon for those with Twitter Blue to be a little nerd emoji wherever it appears, but it only does so when I actively click the normal icon on a profile page, in the little popup that Twitter shows. It works fine in Chrome with Violentmonkey.

Checking the console I get the below error. Adding `@inject-into content` to the script makes it disappear, but the actual issue still occurs regardless.
`[Error] Refused to execute a script because its hash, its nonce, or 'unsafe-inline' does not appear in the script-src directive of the Content Security Policy. (home, line 0)`

I do however also get a bunch of these logs that refer to [this line](https://gist.github.com/busybox11/53c76f57a577a47a19fab649a76f18e3#file-twitterblue-nerd-js-L206) of the script regardless:
`[Log] [nerdtick]
TypeError: undefined is not an object (evaluating 'propsElem.children')
bluetickHandling
handleMutation`

**Steps to reproduce:**
1. Install the script, grant it access to Twitter.com, enable it etc.
2. Load a profile with just Twitter Blue verification [like this one](https://twitter.com/RecklessPatel).
3. See what Verified icon shows up throughout that page, and try clicking the Verified icon on the profile itself to bring up the little popup from Twitter.

**Expected Results:**
The Verified icon is a little yellow nerd emoji in all cases.

**Actual Results:**
The Verified icon is the standard tick, with the exception of if you click the normal icon on the profile page to bring up the popup.

**Comment by quoid** [comment]
I think this issue is more than just `CSP`. 

The script is failing to to get the elements React props in Safari. It relies on the code in this [Stackoverflow answer](https://stackoverflow.com/questions/70507318/how-to-get-react-element-props-from-html-element-with-javascript).

The function `getReactProps` isn't returning anything and the script fails.

@hopkapi this is an issue with the userscript author's code, you should ask them about it

**Comment by trlkly** [comment]
@quoid Would that not be for the same reasons that @ACTCD said? `getReactProps` would rely on the React framework from the page context, which would not be available if you `@inject-into: content`

I did check, and the userscript works just fine on Chrome with TamperMonkey.

**Comment by quoid** [comment]
@trlkly 

Yes, but if you reported it to the author, perhaps they can come up with a different that doesn't rely on those props. The `CSP` issue might not be a total blocker, meaning there is hope a work around could be found.

**Comment by trlkly** [comment]
@quoid 
The only other script I know that handles this sort of thing seems to also use that reactProps code. I suspect it may be the only known way to detect the difference between an old Twitter verified checkmark and the new Twitter Blue verified checkmark. 

Incidentally, this explains the issue here, too: 
https://github.com/wseagar/eight-dollars/issues/19#issuecomment-1312494759

**Comment by quoid** [comment]
@trlkly 

I don't use Twitter, other than getting linked there every once in a while, nor know much about the checkmarks. What is the purpose of these userscripts? 

> eight-dollars can help you tell the difference between actual verified accounts and twitter blue users. 

Unfortunately, I don't know what that means and I suspect I would have to read up on Twitter to figure out the difference. If you know off the top of your head, please do share.

> way to detect the difference between an old Twitter verified checkmark and the new Twitter Blue verified checkmark

I don't know enough about this to even try to offer a solution. What's the motivation for wanting to know the difference and would you ve able to tell is someone had an "old Twitter verified checkmark"?

**Comment by trlkly** [comment]
@quoid

The checkmarks are an indication of account validation. The idea is that Twitter is assuring you that the person behind the account is who they say they are. 

The old system, before Elon Musk bought Twitter, involved actually checking to see if you were who you say you are. The new system, called Twitter Blue, gave you the checkmark if you paid $8. This led to a lot of abuse, with tons of fake accounts. 

A particularly consequential one can be read about here:

https://cardiovascularbusiness.com/topics/healthcare-management/healthcare-economics/elon-musk-and-eli-lilly-how-fake-tweet-caused

----

I'm not sure it can be done. When I look at the HTML source, the checkmarks look identical. But if you want to check, [here is an account that was verified with the old system](https://twitter.com/Twitter/status/1580661436132757506), and [here is one that was verified with the new one](https://twitter.com/anothercohen/status/1593404311832338442).

**Comment by quoid** [comment]
@trlkly awesome, thanks for the summary

> I'm not sure it can be done.

But this can be done checking react props?

**Comment by trlkly** [comment]
@quoid 

It seems like it. I'm not sure it's intentional (by Twitter) that you can tell them apart. I unfortunately don't know how it works exactly. 

---

As for the links, that's quite weird. I copied them from working pages, but, yes, if you click on them, you can't see them. I'll try linking the accounts themselves:

* new verified: https://twitter.com/anothercohen
* old verified: https://twitter.com/Twitter

**Comment by quoid** [comment]
@trlkly 

Thanks

do you know of any other account that are 100% certain to be "new verified"?

**Comment by trlkly** [comment]
@quaid

This one is definitely "new verified"
https://twitter.com/RohunJauhar

This one is detected by the scripts as "new verified":

https://twitter.com/MarcyFoxfire

**Comment by quoid** [comment]
@trlkly 

Thanks for the links and insight. I was hoping to find something outside of those props that could indicate a difference, but like you originally said - it doesn't seem like there is anything that indicates that in the html markup.

Damn, I can see the value in knowing this information.

**Comment by trlkly** [comment]
@quoid Is there no way to bypass CSP and/or modify it with a Safari extension? That sucks. 

BTW, would it be possible for you to release signed binaries that could be used in developer mode in Safari? It seems to be basically impossible to sign in to the App Store using a VM. So the last version I was able to test with was 3.x, not 4.0.

**Comment by quaid** [comment]
> @quaid
> 
> This one is definitely "new verified"
> https://twitter.com/RohunJauhar
> 
> This one is detected by the scripts as "new verified":
> 
> https://twitter.com/MarcyFoxfire
> 

Nice to meet you @quoid :goose: :grin:

Now we know in case this happens again.

**Comment by quoid** [comment]
> > @quaid
> > This one is definitely "new verified"
> > https://twitter.com/RohunJauhar
> > This one is detected by the scripts as "new verified":
> > https://twitter.com/MarcyFoxfire
> 
> Nice to meet you @quoid :goose: 😁
> 
> Now we know in case this happens again.

Ha! 👋

**Comment by trlkly** [comment]
@quoid @quaid 

Sorry. 😅

**Comment by trlkly** [comment]
The latest version of eight-dollars now uses an external list of verified accounts instead of using the window object. Thus the script might actually run. To avoid any other issues, here is a standalone version of the script that contains the list:

https://github.com/trlkly/eight-dollars/raw/main/stand-alone.user.js

And here is the userscript version I've been using on Chrome that tries to stay updated to the official version:
https://github.com/trlkly/eight-dollars/raw/main/script.user.js

**Comment by quoid** [comment]
> // @resource    verified.txt https://raw.githubusercontent.com/wseagar/eight-dollars/main/data/verified.txt

I took a quick look at the user script, assuming that's the link to the external list, it still won't work in the current version due to #224

**Comment by trlkly** [comment]
@quoid Oops. That resource shouldn't actually be necessary in the Stand Alone version, as I completely embedded verified.txt into the code as a template string. I just forgot to edit the manifest. Let me fix that now.
